compiler-core/src/language_server/annotation_printer.rs

```diff
@@ -0,0 +1,218 @@
+#![allow(dead_code)]
+
+use ecow::EcoString;
+use im::HashMap;
+use itertools::Itertools;
+use std::sync::Arc;
+
+use crate::{
+    ast::TypeAst,
+    type_::{Type, TypeVar, PRELUDE_MODULE_NAME},
+};
+
+#[derive(Debug)]
+pub struct AnnotationPrinter<'a> {
+    names: HashMap<u64, EcoString>,
+    uid: u64,
+
+    // A mapping of printd type names to the module that they are defined in.
+    printed_types: HashMap<EcoString, EcoString>,
+
+    // Mapping of module.type to type alias -
+    // eg: import mod1.{type Cat as c} -> Key: (mod1, Cat), Value: c
+    // eg: type UserId = Int -> Key: (mod, UserId), Value: Int
+    type_aliases: &'a HashMap<(EcoString, EcoString), EcoString>,
+
+    // Mapping of module alias to module name
+    module_aliases: &'a HashMap<EcoString, EcoString>,
+    type_parameters: &'a HashMap<u64, TypeAst>,
+    current_module: EcoString,
+}
+
+impl<'a> AnnotationPrinter<'a> {
+    pub fn new(
+        type_aliases: &'a HashMap<(EcoString, EcoString), EcoString>,
+        module_aliases: &'a HashMap<EcoString, EcoString>,
+        type_parameters: &'a HashMap<u64, TypeAst>,
+        current_module: EcoString,
+    ) -> Self {
+        AnnotationPrinter {
+            names: im::HashMap::new(),
+            uid: u64::default(),
+            printed_types: HashMap::new(),
+            type_aliases,
+            module_aliases,
+            type_parameters,
+            current_module,
+        }
+    }
+
+    pub fn print(&mut self, typ: &Type) -> String {
+        let mut typ_str = String::new();
+        match typ {
+            Type::Named {
+                name, args, module, ..
+            } => {
+                let key = (module.clone(), name.clone());
+
+                if let Some(typ_alias) = self.type_aliases.get(&key) {
+                    // Type has been aliased. eg: import mod1.{type Cat as c} or type UserId = Int
+                    if self.name_clashes_if_unqualified(typ_alias, module) {
+                        typ_str.push_str(module);
+                        typ_str.push('.');
+                    }
+                    typ_str.push_str(typ_alias.as_str());
+                } else if module == PRELUDE_MODULE_NAME || *module == self.current_module {
+                    if self.name_clashes_if_unqualified(name, module) {
+                        typ_str.push_str(module);
+                        typ_str.push('.');
+                    }
+                    typ_str.push_str(name.as_str());
+                } else {
+                    // Always qualify types from other modules
```

Comment:
If the type has been imported in an unqualified fashion then we want to use the unqualified constructor, not always qualify it.

Reply:
Oh Yes. Will change this. Might have to keep track of all unqualified imports in the module then.